ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto

```diff
@@ -24,18 +24,84 @@ message TransactionFilter {
   map<string, Filters> filters_by_party = 1;
 }
 
+// Used for filtering Transaction and Active Contract Set streams.
+// Determines which on-ledger events will be served to the client.
+message TransactionExerciseFilter {
+
+  map<string, ExerciseFilters> filters_by_party = 1;
+}
+
+message ExerciseFilters {
+  // Required
+  InclusiveExerciseFilters inclusive = 1;
+}
+
 message Filters {
 
   // If not set, no filters will be applied.
   // Optional
   InclusiveFilters inclusive = 1;
 }
 
+message FilterSelector {
+  oneof {
+    // Include events for this template id.
+    Identifier template_id = 1;
+    // Include events for templates which implement this interface id.
+    Identifier interface_id = 1;
+  }
+}
+
+// The values included for events matching the filter.
+message FilterValue {
+  oneof {
+    // Include create_arguments in CreateEvent
+    boolean create_arguments = 1;
+    // Include create_arguments_blob in CreateEvent
+    boolean create_arguments_blob = 2;
+    // Include interface view if the interface is implemented by the template.
+    InterfaceView interface_view = 3;
```

Comment:
Having thought about this for a bit, we need to enforce something here. We cannot allow completely different interface veiws. My thought initially was that at least superinterfaces are fine and I think that’s still the case but I’m not sure that’s really useful. My suggestion would be that we add the restriction that if you specify an interface view it must either be for the interface or the template must implement the interface. This should also remove the need for the demultiplexing identifiers.

Reply:
> We cannot allow completely different interface veiws.

Why not? If you ask for an interface view that is unrelated to any interface in the selector, you just get no payloads (`CreatedEvent`s with no `interface_views`). Do you just want to prevent obvious user errors, or is there another reason?

Reply:
I might have confused myself before. I was worried that if the selector & the interface view are unrelated, there might be a case retroactive instances would change whether something gets included while a stream is running but thinking about it now I cannot come up with an example.

I think the error case may still be useful though so I’d be inclined to go for the restrictive version first and extend it later if we find a need for it. Removing restrictions is always easier than adding them.